sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,890 @@
+        streamSubPartitioner.releaseBatchesByPartition(curId), spillCallback)
+      val streamCb = withResource(streamSpillBatch) { _ =>
+        streamSpillBatch.map(_.getColumnarBatch())
+      }
+      closeOnExcept(streamCb) { _ =>
+        closeOnExcept(curBuiltBatch.map(_.getColumnarBatch())) { buildCb =>
+          // OK to refresh the build batch and ids
+          curPartIds = curPartIds.tail
+          if (curPartIds.isEmpty) {
+            // Done for the current built batch, close it right away
+            curBuiltBatch.foreach(_.close())
+            curBuiltBatch = None
+          }
+          Some(new BatchPair(buildCb, streamCb))
+        }
+      }
+    } else {
+      // No more data, since the id array returned from "buildSubIterator.next()"
+      // can not be empty.
+      None
+    }
+  }
+
+}
+
+/** Iterators for joins */
+class SubHashExistenceJoinIterator(
+    buildIter: Iterator[ColumnarBatch],
+    val boundBuildKeys: Seq[GpuExpression],
+    buildSchema: StructType,
+    streamIter: Iterator[ColumnarBatch],
+    val boundStreamKeys: Seq[GpuExpression],
+    val boundCondition: Option[GpuExpression],
+    val numFirstConditionTableColumns: Int,
+    val compareNullsEqual: Boolean,
+    numPartitions: Int,
+    spillCallback: SpillCallback,
+    val opTime: GpuMetric,
+    val joinTime: GpuMetric)
+  extends Iterator[ColumnarBatch] with Arm with GpuHashExistenceJoinLike {
+
+  private[this] val batchPairSubIterator = new GpuBatchPairSubPartitionIterator(
+    buildIter, boundBuildKeys, streamIter, boundStreamKeys, numPartitions, spillCallback)
+  use(batchPairSubIterator)
+
+  private[this] var nextBatch: Option[ColumnarBatch] = None
+
+  override def hasNext: Boolean = opTime.ns {
+    if (closed) return false
+    var mayContinue = true
+    while (nextBatch.isEmpty && mayContinue) {
+      if (batchPairSubIterator.hasNext) {
+        nextBatch = joinAndClose(batchPairSubIterator.next())
+      } else {
+        mayContinue = false
+      }
+    }
+    nextBatch.isDefined
+  }
+
+  override def next(): ColumnarBatch = {
+    if (!hasNext) throw new NoSuchElementException()
+    val ret = nextBatch.get
+    nextBatch = None
+    ret
+  }
+
+  override def close(): Unit = {
+    nextBatch.foreach(_.safeClose(new Exception()))
+    nextBatch = None
+    super.close()
+  }
+
+  private[this] def joinAndClose(cbPair: BatchPair): Option[ColumnarBatch] = {
+    withResource(cbPair) { _ =>
+      val (build, stream) = cbPair.get
+      // FIXME what's expected if getting at least one empty batch for existence join?
+      // Now return None if stream batch is empty, since existence join always uses
+      // stream as the left side.
+      stream.map { streamBatch =>
+        val buildBatch = build.getOrElse(GpuColumnVector.emptyBatch(buildSchema))
+        existenceJoinNextBatch(streamBatch, buildBatch)
+      }
+    }
+  }
+}
+
+abstract class BaseSubHashJoinIterator(
```

Comment:
My main concern with this is the new Sub variants of the join classes. I wish that either we remove the old ones or retrofit them to be able to handle the sub partitioning. Having alternate implementations seems like something we would like to avoid.

Reply:
Thanks, it makes me re-think about some framework things.

Reply:
Updated to leverage the original join iterators by calling `doJoin` inside this sub join iterator, suggested by Bobby. And now this sub iterator looks like a high order iterator taking care of sub-partitioning things.